README.md
[![Build Status](https://travis-ci.org/express42-cookbooks/openvpn.svg?branch=master)](https://travis-ci.org/express42-cookbooks/openvpn)

# Description

Installs and configures OpenVPN.

# Requirements

## Platform:

* Ubuntu
* RHEL
* CentOS

## Gems:

* knife-openvpn

# Attributes

## Default

* `node['openvpn']['server_name']` -  Defaults to `"default"`.
* `node['openvpn']['install_epel']` -  Defaults to `true`.
* `node['openvpn']['default']['remote_host']` -  Defaults to `"vpn.example.com"`.
* `node['openvpn']['default']['server_ip']` -  Defaults to `"127.0.0.1"`.
* `node['openvpn']['default']['port']` -  Defaults to `"1194"`.
* `node['openvpn']['default']['proto']` -  Defaults to `"udp"`.
* `node['openvpn']['default']['dev']` -  Defaults to `"tun"`.
* `node['openvpn']['default']['mode']` -  Defaults to `"routed"`.
* `node['openvpn']['default']['netmask']` -  Defaults to `"255.255.255.0"`.
* `node['openvpn']['default']['subnet']` -  Defaults to `"127.0.1.0"`.
* `node['openvpn']['default']['network_bridge']` -  Defaults to `"br0"`.
* `node['openvpn']['default']['network_interface']` -  Defaults to `"eth0"`.
* `node['openvpn']['default']['dhcp_start']` -  Defaults to `"127.0.0.100"`.
* `node['openvpn']['default']['dhcp_end']` -  Defaults to `"127.0.0.150"`.
* `node['openvpn']['default']['verb']` -  Defaults to `"3"`.
* `node['openvpn']['default']['push']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['duplicate_cn']` -  Defaults to `"false"`.
* `node['openvpn']['default']['client_to_client']` -  Defaults to `"false"`.
* `node['openvpn']['default']['keepalive_interval']` -  Defaults to `"10"`.
* `node['openvpn']['default']['keepalive_timeout']` -  Defaults to `"60"`.
* `node['openvpn']['default']['comp_lzo']` -  Defaults to `"true"`.
* `node['openvpn']['default']['link_mtu']` -  Defaults to `"nil"`.
* `node['openvpn']['default']['tun_mtu']` -  Defaults to `"nil"`.
* `node['openvpn']['default']['cipher']` -  Defaults to `"false"`.
* `node['openvpn']['default']['redirect_gateway']` -  Defaults to `"false"`.
* `node['openvpn']['default']['push_dns_server']` -  Defaults to `"false"`.
* `node['openvpn']['default']['script_security']` -  Defaults to `"1"`.
* `node['openvpn']['default']['use_tls_auth']` -  Defaults to `"true"`.
* `node['openvpn']['default']['chroot']` -  Defaults to `"false"`.
* `node['openvpn']['default']['client_config_dir']` -  Defaults to `"false"`.
* `node['openvpn']['default']['ccd_exclusive']` -  Defaults to `"false"`.
* `node['openvpn']['default']['users']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['revoked_users']` -  Defaults to `"[ ... ]"`.
* `node['openvpn']['default']['ifconfig_pool_persist']` - Defaults to `"true"`.
* `node['openvpn']['client']['remote_servers']` - Defaults to `"[ ... ]"`.

## Iptables

* `node['openvpn']['iptables']['postrouting']` -  Defaults to `true`.
* `node['openvpn']['iptables']['interface']` -  Defaults to `eth0`.

## Sysctl

* `node['openvpn']['ip_forward']` -  Defaults to `true`.

# Recipes

* openvpn::default - Installs and configures OpenVPN.
* openvpn::sysctl - Configures IP forwarding via sysctl
* openvpn::iptables - Configures postrouting via iptables
* openvpn::client - Configures client connection to server

# Server Modes

* Routed

For routed network you must define vpn ```subnet```, like in previous example

* Bridged

Bridged setup need more configuration and configured network bridge on your server

```
"default_attributes": {
  "openvpn": {
    "server_name": "office",
    "office": {
      "remote_host": "vpn.example.com",
      "server_ip": "10.90.5.5",
      "port": "443",
      "proto": "tcp",
      "dev": "tap",
      "verb": "3",
      "mode": "bridged",
      "script_security": "2",
      "dhcp_start": "10.90.5.100",
      "dhcp_end": "10.90.5.240",
      "network_bridge": "br0",
      "network_interface": "eth0"
    }
  }
}

```

See fixture cookbook in `tests/fixtures/cookbooks`.


# To-do

1. Revoke access
2. Import existing certs/keys
3. Add support for client recipe-friendly config generation


# Usage

For example you want to setup vpn server and call it ```office```

* Ensure that you have ```.chef/encrypted_data_bag_secret```.
Otherwise you can generate one with ```openssl rand -base64 512 > .chef/encrypted_data_bag_secret```

* Install knife plugin:

  ```
  gem install knife-openvpn
  ```

* Create server certificate authority, server cert/key, DH params:

  ```
  knife openvpn server create office
  ```

  ```office``` - is a name of vpn-server, there is some limitations on this: no dots, no commas, no spaces, no special symbols for reasons.

* Great, now check ```data_bags``` directory, you will find new databag ```openvpn-office``` with few items for ca, dh, cert/key pair and some openssl config. Now it is time to upload it to Chef server:

  ```
  knife data bag create openvpn-office --secret-file=.chef/encrypted_data_bag_secret
  knife data bag from file openvpn-office data_bags/openvpn-office/*
  ```

* Add ```recipe[openvpn]``` to node run_list, and override default attributes:

  ```
  "run_list": [
  "recipe[openvpn]"
  ],
  "default_attributes": {
    "openvpn": {
      "server_name": "office",
      "office": {
        "remote_host": "vpn.example.com",
        "server_ip": "10.90.5.5",
        "subnet": "10.200.1.0",
        "port": "443",
        "proto": "tcp",
        "dev": "tun",
        "verb": "3",
        "push": [
        "route 10.90.0.0 255.255.255.0",
        "route 10.90.1.0 255.255.255.0"
        ]
      }
    }
  }

  ```

* Add ```recipe[openvpn::sysctl]``` if you need to setup net.ipv4.ip_forward with this cookbook. ```node['openvpn']['ip_forward']``` should be set to ```true``` (it's ```true``` by default).

* Add ```recipe[openvpn::iptables]``` if you need to setup nat postrouting with this cookbook.

  Chef, run!

* When server is up and running we can add some users to start use it.
No moar certificate management pain:

  ```
  knife openvpn user create office john
  knife data bag from file openvpn-office data_bags/openvpn-office/john.json
  ```

* Export vpn-client data and send it to John:

  ```
  knife openvpn user export office john
  ```
resulting archive contains config (.ovpn), ca cert, John's cert and key

* Revokation of user certificate is also possible:
  ```
  knife openvpn user revoke office john
  knife data bag from file openvpn-office data_bags/openvpn-office/openvpn-crl.json
  ```

# Client configuration

* Add ```recipe[openvpn::client]``` to run_list

* Add data bag item for each server in ```node['openvpn']['client']['remote_servers']``` containing next elements (replace new lines with '\n'):

  ```
  "ca" - contents of ca.crt generated with knife-openvpn
  "crt" - contents of client's certificate
  "key" - contents of client's private key
  "conf" - contents of client's configuration
  ```

# License and Maintainer

Maintainer:: LLC Express 42 (<cookbooks@express42.com>)

License:: MIT


CHANGELOG.md
# CHANGELOG

This file is used to list changes made in each version of the openvpn cookbook.

0.1.7
-----
- Added support for client configuration

0.1.6
-----
- New attribute ifconfig_pool_persist (bool)

0.1.5
-----
- Recreate pem files if they where changed
- Restart openvpn service if server files were changed


0.1.4
-----
- Setup iptables and sysctl
- Initial RHEL support

0.1.3
-----
- Remove knife plugin

0.1.2
-----
- Fix Rubocop warnings
- Add Test Kitchen, Rubocop, Berkshelf, Serverspec and Travis CI
- Add encrypted data bags to integration test
- Update docs

0.1.1
-----
- Fix permission for crl file

0.1.0
-----
- Initial release of openvpn

- - -
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


metadata.rb
name             'openvpn'
maintainer       'LLC Express 42'
maintainer_email 'cookbooks@express42.com'
license          'MIT'
description      'Installs and configures OpenVPN.'
long_description IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version          '0.1.7'
source_url 'https://github.com/express42-cookbooks/openvpn' if respond_to?(:source_url)
issues_url 'https://github.com/express42-cookbooks/openvpn/issues' if respond_to?(:issues_url)

recipe 'openvpn::default', 'Installs and configures OpenVPN.'
recipe 'openvpn::sysctl', 'Configures IP forwarding via sysctl.'
recipe 'openvpn::iptables', 'Configures postrouting via iptables.'

%w(
  amazon
  centos
  debian
  fedora
  redhat
  scientific
  ubuntu
).each do |os|
  supports os
end

depends 'iptables'
depends 'sysctl'
depends 'yum-epel'


recipes/client.rb
#
# Cookbook Name:: openvpn
# Recipe:: client

# Copyright 2016, LLC Express 42
#
# Permission is hereby granted, free of charge, to any person obtaining
# a copy of this software and associated documentation files (the
# "Software"), to deal in the Software without restriction, including
# without limitation the rights to use, copy, modify, merge, publish,
# distribute, sublicense, and/or sell copies of the Software, and to
# permit persons to whom the Software is furnished to do so, subject to
# the following conditions:
#
# The above copyright notice and this permission notice shall be
# included in all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
#

if node['platform_family'] == 'rhel' && node['openvpn']['install_epel']
  include_recipe 'yum-epel'
end

package 'openvpn'

user 'openvpn'
group 'openvpn' do
  members ['openvpn']
end

directory '/etc/openvpn/' do
  owner 'root'
  group 'openvpn'
  mode '0770'
end

directory '/var/log/openvpn' do
  owner 'root'
  group 'root'
  mode '0755'
end

node['openvpn']['client']['remote_servers'].each do |server_name|
  client_item = Chef::EncryptedDataBagItem.load('openvpn-client', server_name.to_s)

  files = {
    "#{server_name}-ca.crt" => client_item['ca'],
    "#{server_name}.crt" => client_item['crt'],
    "#{server_name}.key" => client_item['key'],
    "#{server_name}.conf" => client_item['conf']
  }

  service_name = node['platform_family'] == 'rhel' && node['platform_version'].to_f >= 7.0 ? "openvpn@#{server_name}" : 'openvpn'

  files.each do |name, content|
    file "/etc/openvpn/#{name}" do
      owner 'openvpn'
      group 'openvpn'
      mode '0600'
      content content
      sensitive true
      notifies :restart, "service[#{service_name}]", :delayed
    end
  end

  # needed due to SystemD before version 208-20.el7_1.5 not supporting enable for @ services
  # https://bugzilla.redhat.com/show_bug.cgi?id=1142369
  link "/etc/systemd/system/multi-user.target.wants/#{service_name}.service" do
    to '/usr/lib/systemd/system/openvpn@.service'
    link_type :symbolic
    not_if { service_name == 'openvpn' }
  end

  service service_name do
    action [:start]
  end

  service service_name do
    action [:enable]
    only_if { service_name == 'openvpn' }
  end
end


attributes/default.rb
# Name of openvpn server.
default['openvpn']['server_name'] = 'default'

# Install epel on RHEL based hosts
default['openvpn']['install_epel'] = true

# Set net.ipv4.ip_forward to 1
default['openvpn']['ip_forward'] = true

default['openvpn']['iptables']['postrouting'] = true

# External interface VPN traffic will go out of to the outside world
default['openvpn']['iptables']['interface'] = 'eth0'

# This address will be used for clients config as vpn server address.
default['openvpn']['default']['remote_host'] = 'vpn.example.com'

# Server is accepting connections on this IP.
default['openvpn']['default']['server_ip'] = '127.0.0.1'
# Network port to listen on.
default['openvpn']['default']['port'] = '1194'
# Protocol to use (tcp or udp).
default['openvpn']['default']['proto'] = 'udp'
# Type of dev to use (tun or tap).
default['openvpn']['default']['dev'] = 'tun'
# OpenVPN server mode: routed or bridged.
default['openvpn']['default']['mode'] = 'routed'
# Netmask to use (for routed and bridged modes).
default['openvpn']['default']['netmask'] = '255.255.255.0'
# Only in routed mode. This subnet is used for vpn client addresses. Server takes first address in this subnet.
default['openvpn']['default']['subnet'] = '127.0.1.0'
# Only in bridged mode. OpenVPN tap interface will be added to this bridge (see up/down scripts).
default['openvpn']['default']['network_bridge'] = 'br0'
# Only in bridged mode. Network interface used for bridging, we need to turn promisc mode for it.
default['openvpn']['default']['network_interface'] = 'eth0'
# Only in bridged mode. Assign client addresses from this range.
default['openvpn']['default']['dhcp_start'] = '127.0.0.100'
default['openvpn']['default']['dhcp_end'] = '127.0.0.150'
default['openvpn']['default']['verb'] = '3'

# Array of route strings that will be pushed on client connect.
default['openvpn']['default']['push'] = []
default['openvpn']['default']['duplicate_cn'] = false
default['openvpn']['default']['ifconfig_pool_persist'] = true
# True\False. Allow clients talk to each other.
default['openvpn']['default']['client_to_client'] = false
default['openvpn']['default']['keepalive_interval'] = 10
default['openvpn']['default']['keepalive_timeout'] = 60
# True\False. Enable compression.
default['openvpn']['default']['comp_lzo'] = true
default['openvpn']['default']['link_mtu'] = nil
default['openvpn']['default']['tun_mtu'] = nil
default['openvpn']['default']['cipher'] = false
# True\False. Send all traffic through vpn channel
default['openvpn']['default']['redirect_gateway'] = false
default['openvpn']['default']['push_dns_server'] = false
# Set it to "2" for bridged mode to allow script execution needed to configure network interfaces
default['openvpn']['default']['script_security'] = 1
# Use tls-auth
default['openvpn']['default']['use_tls_auth'] = true
# Use a chroot environment
default['openvpn']['default']['chroot'] = false
# Setup a client config directory
default['openvpn']['default']['client_config_dir'] = false
# Force each client to have a ccd config file (ignored if client_config_dir is false)
default['openvpn']['default']['ccd_exclusive'] = false

default['openvpn']['default']['users'] = []
default['openvpn']['default']['revoked_users'] = []

# Names of remote servers for client configuration
default['openvpn']['client']['remote_servers'] = []
